atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientFactoryImpl.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    private static final class CqlResourceHandleImpl implements CqlResourceHandle {
+
+        private static final class InitializingWrapper extends AsyncInitializer
+                implements AutoDelegate_CqlResourceHandle {
+
+            private volatile CqlResourceHandle cqlResourceHandle;
+            private final Cluster cluster;
+            private final ExecutorService executorService;
+
+            InitializingWrapper(ExecutorService executorService, Cluster cluster) {
+                this.executorService = executorService;
+                this.cluster = cluster;
+            }
+
+            @Override
+            protected void tryInitialize() {
+                cqlResourceHandle = new CqlResourceHandleImpl(executorService, cluster.connect());
+            }
+
+            @Override
+            protected String getInitializingClassName() {
+                return "CqlResourceHandle";
+            }
+
+            @Override
+            public CqlResourceHandle delegate() {
+                isInitialized();
+                return cqlResourceHandle;
+            }
+        }
+
+        private static final Logger log = LoggerFactory.getLogger(CqlResourceHandleImpl.class);
+
+        private final ExecutorService executorService;
+        private final Session session;
+
+        static CqlResourceHandle create(ExecutorService executorService, Cluster cluster, boolean initializeAsync) {
+            if (initializeAsync) {
+                return new InitializingWrapper(executorService, cluster);
+            }
+            return new CqlResourceHandleImpl(executorService, cluster.connect());
+        }
+
+        private CqlResourceHandleImpl(ExecutorService executorService, Session session) {
+            this.executorService = executorService;
+            this.session = session;
+        }
+
+        @Override
+        public Session session() {
+            return session;
+        }
+
+        @Override
+        public Executor executor() {
+            return executorService;
+        }
+
+        @Override
+        public void close() {
+            try {
+                executorService.shutdown();
+                if (executorService.awaitTermination(10, TimeUnit.SECONDS)) {
+                    log.info("Executor service terminated properly.");
+                } else {
+                    log.warn("Executor service timed out before shutting down, shutting down forcefully");
+                    executorService.shutdownNow();
+                }
+            } catch (InterruptedException e) {
+                log.warn("Thread was interrupted while waiting for executor service to terminate.", e);
+            } catch (Exception e) {
+                log.warn("Exception on executor service termination", e);
+            }
+            Cluster cluster = session.getCluster();
+            session.close();
+            cluster.close();
```

Comment:
this isn't really what you want, because if you close this `cluster`, then every other session that is created from this cluster then gets shutdown.

Reply:
My idea was to make this factory "stupid" so this was intentional. However your comments do make sense.

Reply:
Implemented caching resource manager.